dyn-cache/src/namespace.rs

```diff
@@ -100,23 +115,37 @@ where
         Input: Borrow<Arg>,
     {
         let hashed = self.hashed(key);
-        self.entry(&hashed)
-            .and_then(|(_, cell)| cell.get(input))
-            .ok_or_else(|| KeyMiss::hashed(hashed))
+        if let Some((_, cell)) = self.entry(&hashed) {
+            cell.get(input, dependent).map_err(|d| KeyMiss::hashed(hashed, None, d))
+        } else {
+            let node = DepNode::new();
+            node.root(dependent);
+            let new_dep = node.as_dependent();
+            Err(KeyMiss::hashed(hashed, Some(node), new_dep))
+        }
     }
 
-    pub fn store<Key>(&mut self, hashed: KeyMiss<'_, Key, H>, input: Input, output: Output)
+    pub fn store<Key>(&mut self, miss: KeyMiss<'_, Key, H>, input: Input, output: Output)
     where
         Key: Eq + Hash + ToOwned<Owned = Scope> + ?Sized,
         Scope: Borrow<Key>,
     {
-        let hashed = hashed.inner.unwrap_or_else(|k| self.hashed(k));
+        let dependent = miss.dependent;
+        let hashed = miss.inner.unwrap_or_else(|k| self.hashed(k));
         match self.entry_mut(&hashed) {
             RawEntryMut::Occupied(occ) => {
-                occ.into_mut().store(input, output);
+                debug_assert!(miss.node.is_none(), "mustn't create nodes that aren't used");
```

Comment:
I don't think so, but I wonder if there's a way to make this a static invariant.